test/opaque_closure.jl

```diff
@@ -230,3 +230,26 @@ call_global_opaque_closure() = GLOBAL_OPAQUE_CLOSURE()
 let oc = @opaque a->sin(a)
     @test length(code_typed(oc, (Int,))) == 1
 end
+
+# constructing an opaque closure from IRCode
+using Core.Compiler: IRCode
+using Core: CodeInfo
+
+function OC(ir::IRCode, env...)
+    src = ccall(:jl_new_code_info_uninit, Ref{CodeInfo}, ());
+    src.slotflags = UInt8[]
+    nargs = length(ir.argtypes)
+    src.slotnames = fill(:none, nargs)
+    Core.Compiler.replace_code_newstyle!(src, ir, nargs);
+    Core.Compiler.widen_all_consts!(src);
+    src.inferred = true
+    # NOTE: we need ir.argtypes[1] == typeof(env)
+
+    ccall(:jl_new_opaque_closure_from_code_info, Any, (Any, Any, Any, Any, Any, Cint, Any, Cint, Cint, Any),
+          Tuple{ir.argtypes[2:end]...}, Union{}, Any, @__MODULE__, src, 0, nothing, nargs-1, false, env)
```

Comment:
IIRC we had said that we did want to separate the method construction from the actual opaque closure creation. I don't need it immediately, but I think it would be perfectly sensible to construct several opaque closures with the same inferred code by different `env`s (after all, otherwise, you could have just constproped `env` at inference time.

Reply:
Agreed, just starting with the minimal API surface.

Reply:
Actually it's a bit tricky, since with this there are two kinds of methods, one that can be specialized as normal, and one that is already inferred. If the same OpaqueClosure constructor can accept both, it needs some additional checks.

Reply:
I think we are always permitted to specialize methods in order to do non-observable optimizations (e.g. the normal inference pipeline) on anything we compile, or we couldn't really pass them to codegen either (which also runs various compilation and inference stages). Only the other syntax form has the ability to create partial specializations at a call-site and optimize the env on those (while preserving the opaque characteristic).